File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {

        int n = s.length(); 
        unordered_map<char, int>frequency;
        int i = 0; int result = INT_MIN;
        for(int j = 0; j < n; j++){
            frequency[s[j]]++;
            while(i < j && frequency[s[j]] > 1){
                frequency[s[i]]--;
                i++;
            }
            result = max(result, j - i + 1);
        }

        return result == INT_MIN ? 0 : result;
    }
};

File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        int n = s.length(); 
        Map<Character, Integer> map = new HashMap<>();
        int start = 0; int len = 0;
        for(int end = 0; end < n; end++){
            char c = s.charAt(end);
            if(map.containsKey(c)){
                start = Math.max(map.get(c) + 1, start);
            }

            len = Math.max(len, end - start + 1);
            map.put(c, end);
        }

        return len;
    }
}